lib/membrane/video_compositor.ex

```diff
@@ -108,155 +112,49 @@ defmodule Membrane.VideoCompositor do
-    end)
-    |> Map.new()
-  end
-
-  defp remove_finished_tracks(state) do
-    state.ids_to_tracks
-    |> Enum.reduce(
-      state,
-      fn {id, %Track{} = track}, state ->
-        if Track.finished?(track) do
-          remove_track(state, id)
-        else
-          state
-        end
-      end
-    )
-  end
-
-  defp track_finished?(ids_to_tracks, id) do
-    track = Map.get(ids_to_tracks, id)
-    Track.finished?(track)
-  end
-
-  defp remove_track(state, id) do
-    %{ids_to_tracks: ids_to_tracks, internal_state: internal_state} = state
-    ids_to_tracks = Map.delete(ids_to_tracks, id)
-    {:ok, internal_state} = state.compositor_module.remove_video(internal_state, id)
-    %{state | ids_to_tracks: ids_to_tracks, internal_state: internal_state}
-  end
-
-  defp update_track_status(ids_to_tracks, id, status) do
-    ids_to_tracks
-    |> Map.update!(id, fn track -> %Track{track | status: status} end)
-  end
+    id = Map.get(pads_to_ids, pad)
 
-  defp tracks_status(ids_to_tracks) do
-    if Map.values(ids_to_tracks) |> Enum.all?(&Track.finished?/1) do
-      :all_finished
-    else
-      :still_ongoing
+    %Membrane.Buffer{payload: frame, pts: pts} = buffer
+
+    case state.compositor_module.upload_frame(internal_state, id, {frame, pts}) do
+      {{:ok, {frame, pts}}, internal_state} ->
+        {
+          {
+            :ok,
+            buffer: {
+              :output,
+              %Membrane.Buffer{payload: frame, pts: pts}
+            }
+          },
+          %{state | internal_state: internal_state}
+        }
+
+      {:ok, internal_state} ->
+        {:ok, %{state | internal_state: internal_state}}
     end
   end
 
   @impl true
   def handle_end_of_stream(
         pad,
-        _context,
+        context,
         state
       ) do
-    %{ids_to_tracks: ids_to_tracks, pads_to_ids: pads_to_ids} = state
+    %{pads_to_ids: pads_to_ids, internal_state: internal_state} = state
     id = Map.get(pads_to_ids, pad)
 
-    ids_to_tracks = ids_to_tracks |> update_track_status(id, :end_of_stream)
-    state = %{state | ids_to_tracks: ids_to_tracks}
-
-    state =
-      if track_finished?(ids_to_tracks, id) do
-        remove_track(state, id)
-      else
-        state
-      end
-
-    {{buffers_status, buffers}, state} = merge_frames(state)
-
-    case {buffers_status, tracks_status(state.ids_to_tracks)} do
-      {:empty, :still_ongoing} ->
-        {:ok, state}
-
-      {:empty, :all_finished} ->
-        {{:ok, end_of_stream: :output}, state}
-
-      {:merged, :still_ongoing} ->
-        {{:ok, buffer: {:output, buffers}}, state}
+    {:ok, internal_state} = state.compositor_module.send_end_of_stream(internal_state, id)
+    state = %{state | internal_state: internal_state}
 
-      {:merged, :all_finished} ->
-        {{:ok, buffer: {:output, buffers}, end_of_stream: :output}, state}
+    if Map.to_list(context.pads)
+       |> Enum.all?(fn {ref, pad} -> ref == :output or pad.end_of_stream? end) do
```

Comment:
Move this check to a separate function

The well-named function is more readable and better shows your intentions